In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns

In [2]:
df = pd.read_csv('mushrooms.csv')
df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

In [4]:
for col in df.columns:
    print(df[col].value_counts())

e    4208
p    3916
Name: class, dtype: int64
x    3656
f    3152
k     828
b     452
s      32
c       4
Name: cap-shape, dtype: int64
y    3244
s    2556
f    2320
g       4
Name: cap-surface, dtype: int64
n    2284
g    1840
e    1500
y    1072
w    1040
b     168
p     144
c      44
r      16
u      16
Name: cap-color, dtype: int64
f    4748
t    3376
Name: bruises, dtype: int64
n    3528
f    2160
s     576
y     576
a     400
l     400
p     256
c     192
m      36
Name: odor, dtype: int64
f    7914
a     210
Name: gill-attachment, dtype: int64
c    6812
w    1312
Name: gill-spacing, dtype: int64
b    5612
n    2512
Name: gill-size, dtype: int64
b    1728
p    1492
w    1202
n    1048
g     752
h     732
u     492
k     408
e      96
y      86
o      64
r      24
Name: gill-color, dtype: int64
t    4608
e    3516
Name: stalk-shape, dtype: int64
b    3776
?    2480
e    1120
c     556
r     192
Name: stalk-root, dtype: int64
s    5176
k    2372
f     552
y      24
Name: stalk-surf

All are categorical features and are nominal

### Checking for Class Imbalance

In [5]:
df['class'].value_counts()

e    4208
p    3916
Name: class, dtype: int64

Class is balances since there is not a huge difference in the counts in each class

### Renaming and replacing Target variable

In [6]:
mapping = {"e": 1, "p": 0}

df.rename({'class':'edible'}, axis=1, inplace=True)
df['edible'] = df['edible'].replace(mapping)

In [7]:
df.head()

edible cap-shape cap-surface cap-color bruises odor gill-attachment  \
0       0         x           s         n       t    p               f   
1       1         x           s         y       t    a               f   
2       1         b           s         w       t    l               f   
3       0         x           y         w       t    p               f   
4       1         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

### Create X and y variables

In [8]:
X = df.drop('edible', axis=1)
y = df['edible']

### Convert object to integers

In [9]:
# OneHotEncoding to X
X = pd.get_dummies(X, drop_first=True )
X.head()

cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  cap-shape_x  \
0            0            0            0            0            1   
1            0            0            0            0            1   
2            0            0            0            0            0   
3            0            0            0            0            1   
4            0            0            0            0            1   

   cap-surface_g  cap-surface_s  cap-surface_y  cap-color_c  cap-color_e  ...  \
0              0              1              0            0            0  ...   
1              0              1              0            0            0  ...   
2              0              1              0            0            0  ...   
3              0              0              1            0            0  ...   
4              0              1              0            0            0  ...   

   population_n  population_s  population_v  population_y  habitat_g  \
0             0             1             0             0          0   
1             1             0             0             0          1   
2             1             0             0             0          0   
3             0             1             0             0          0   
4             0             0             0             0          1   

   habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0          0          0          0          1          0  
1          0          0          0          0          0  
2          0          1          0          0          0  
3          0          0          0          1          0  
4          0          0          0          0          0  

[5 rows x 95 columns]

In [10]:
df.head()

edible cap-shape cap-surface cap-color bruises odor gill-attachment  \
0       0         x           s         n       t    p               f   
1       1         x           s         y       t    a               f   
2       1         b           s         w       t    l               f   
3       0         x           y         w       t    p               f   
4       1         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [11]:
df

edible cap-shape cap-surface cap-color bruises odor gill-attachment  \
0          0         x           s         n       t    p               f   
1          1         x           s         y       t    a               f   
2          1         b           s         w       t    l               f   
3          0         x           y         w       t    p               f   
4          1         x           s         g       f    n               f   
...      ...       ...         ...       ...     ...  ...             ...   
8119       1         k           s         n       f    n               a   
8120       1         x           s         n       f    n               a   
8121       1         f           s         n       f    n               a   
8122       0         k           y         n       f    y               f   
8123       1         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

## Create training and testing data

In [12]:
from sklearn.model_selection import train_test_split

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state = 42)

In [14]:
X_train.shape, X_test.shape

((5443, 95), (2681, 95))

### Decision Tree Classifier with criterion entropy

In [15]:
from sklearn.tree import DecisionTreeClassifier

clf_en = DecisionTreeClassifier( criterion='entropy', max_depth = 3, random_state = 0)

clf_en.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=3, random_state=0)

#### Predict the Test set results with criterion entropy

In [16]:
y_pred_en = clf_en.predict(X_test)

#### Check accuracy score with criterion entropy

In [17]:
from sklearn.metrics import accuracy_score

print('Model accuracy score with criterion entropy: {0:0.4f}'. format(accuracy_score(y_test, y_pred_en)))

Model accuracy score with criterion entropy: 0.9571


#### Compare the train-set and test-set accuracy

In [18]:
y_pred_train_en = clf_en.predict(X_train)

print('Training-set accuracy score: {0:0.4f}'. format(accuracy_score(y_train, y_pred_train_en)))

Training-set accuracy score: 0.9638


The training-set accuracy score is 0.9638 while the test-set accuracy to be  0.9571. 
<br> These two values are quite comparable. So, there is no sign of overfitting.

## Confusion Matrix

In [19]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred_en)

print('Confusion matrix\n\n', cm)

Confusion matrix

 [[1300    3]
 [ 112 1266]]


## Classification Report

In [20]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred_en))

              precision    recall  f1-score   support

           0       0.92      1.00      0.96      1303
           1       1.00      0.92      0.96      1378

    accuracy                           0.96      2681
   macro avg       0.96      0.96      0.96      2681
weighted avg       0.96      0.96      0.96      2681



# Random Forest Classifier

### Random Forest Classifier model with default parameters

In [21]:
from sklearn.ensemble import RandomForestClassifier

rfc = RandomForestClassifier(random_state=0)

In [22]:
rfc.fit(X_train, y_train)

y_pred = rfc.predict(X_test)

In [23]:
from sklearn.metrics import accuracy_score

print('Model accuracy score with 10 decision-trees : {0:0.4f}'. format(accuracy_score(y_test, y_pred)))

Model accuracy score with 10 decision-trees : 1.0000


In [30]:
y_pred_train = rfc.predict(X_train)

print('Training-set accuracy score: {0:0.4f}'. format(accuracy_score(y_train, y_pred_train)))

Training-set accuracy score: 1.0000


### Random Forest Classifier model with parameter n_estimators=100

In [24]:
rfc_100 = RandomForestClassifier(n_estimators= 100, random_state=0)

rfc_100.fit(X_train, y_train)

y_pred_100 = rfc.predict(X_test)

In [25]:
print('Model accuracy score with 100 decision-trees : {0:0.4f}'. format(accuracy_score(y_test, y_pred_100)))

Model accuracy score with 100 decision-trees : 1.0000


In [29]:
y_pred_train = rfc_100.predict(X_train)

print('Training-set accuracy score: {0:0.4f}'. format(accuracy_score(y_train, y_pred_train)))

Training-set accuracy score: 1.0000


###  Find important features with Random Forest model

In [27]:
clf = RandomForestClassifier(n_estimators= 100, random_state=0)

clf.fit(X_train, y_train)

RandomForestClassifier(random_state=0)

In [28]:
# view feature scores

feature_scores = pd.Series(clf.feature_importances_,index=X_train.columns).sort_values(ascending=False)
feature_scores

odor_n                      0.130215
gill-size_n                 0.104339
odor_f                      0.085203
ring-type_p                 0.052022
spore-print-color_h         0.046557
                              ...   
stalk-color-above-ring_y    0.000162
cap-color_r                 0.000106
cap-shape_c                 0.000070
spore-print-color_y         0.000000
spore-print-color_o         0.000000
Length: 95, dtype: float64